core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -1193,18 +1202,22 @@ private void scheduleTasks()
         {
             long standardTasksWaitingForNode = getWaitingForNodeTasksCount(STANDARD);
             long speculativeTasksWaitingForNode = getWaitingForNodeTasksCount(SPECULATIVE);
+            long eagerSpeculativeTasksWaitingForNode = getWaitingForNodeTasksCount(EAGER_SPECULATIVE);
 
             while (!schedulingQueue.isEmpty()) {
                 PrioritizedScheduledTask scheduledTask;
-                if (schedulingQueue.getTaskCount(STANDARD) > 0) {
+
+                if (schedulingQueue.getTaskCount(EAGER_SPECULATIVE) > 0 && eagerSpeculativeTasksWaitingForNode < maxTasksWaitingForNode) {
```

Comment:
Should `maxTasksWaitingForNode` account for both `standard` and `eagerSpeculative`?

Reply:
I think it does not matter much - it is mostly about how much state we want to keep. The waiting tasks are not really utilizing resources.
How I handle `EAGER_SPECULATIVE` matches logic we had so far for `STANDARD` and `SPECULATIVE` which are accounted separately.
Having separate lists may be better in a way if `EAGER_SPECULATIVE` require less resources than `STANDARD` task. Then we have a chance to schedule it even if bigger tasks are waiting.